pooch/core.py

```diff
@@ -669,23 +698,42 @@ def download_action(path, known_hash):
     return action, verb
 
 
-def stream_download(url, fname, known_hash, downloader, pooch=None):
+def stream_download(url, fname, known_hash, downloader, pooch=None, retry_if_failed=0):
     """
     Stream the file and check that its hash matches the known one.
 
     The file is first downloaded to a temporary file name in the cache folder.
     It will be moved to the desired file name only if the hash matches the
     known hash. Otherwise, the temporary file is deleted.
 
+    If the download fails for either a bad connection or a hash mismatch, we
+    will retry the download the specified number of times in case the failure
+    was due to a network error.
     """
     # Ensure the parent directory exists in case the file is in a subdirectory.
     # Otherwise, move will cause an error.
     if not fname.parent.exists():
         os.makedirs(str(fname.parent))
-
-    # Stream the file to a temporary so that we can safely check its hash
-    # before overwriting the original.
-    with temporary_file(path=str(fname.parent)) as tmp:
-        downloader(url, tmp, pooch)
-        hash_matches(tmp, known_hash, strict=True, source=str(fname.name))
-        shutil.move(tmp, str(fname))
+    download_attempts = 1 + retry_if_failed
+    max_wait = 10
+    for i in range(download_attempts):
+        try:
+            # Stream the file to a temporary so that we can safely check its
+            # hash before overwriting the original.
+            with temporary_file(path=str(fname.parent)) as tmp:
+                downloader(url, tmp, pooch)
+                hash_matches(tmp, known_hash, strict=True, source=str(fname.name))
+                shutil.move(tmp, str(fname))
+            break
+        except (ValueError, requests.exceptions.RequestException):
+            if i == download_attempts - 1:
+                raise
+            retries_left = download_attempts - (i + 1)
+            get_logger().info(
+                "Failed to download '%s'. "
+                "Will attempt the download again %d more time%s.",
+                str(fname.name),
+                retries_left,
+                "s" if retries_left > 1 else "",
+            )
+            time.sleep(i + 1 if i < max_wait else max_wait)
```

Comment:
Same as this?

```suggestion
            time.sleep(min(i + 1, max_wait))
```

Reply:
I was thinking to myself: I know there is a better way to do this but can't think of it right now 🙂